ethexe/consensus/src/validator/subordinate.rs

```diff
@@ -0,0 +1,399 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2025 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::{initial::Initial, DefaultProcessing, PendingEvent, StateHandler, ValidatorContext};
+use crate::{validator::participant::Participant, ConsensusEvent};
+use anyhow::Result;
+use derive_more::{Debug, Display};
+use ethexe_common::{ProducerBlock, SimpleBlockData};
+use ethexe_signer::{Address, SignedData};
+use gprimitives::H256;
+use std::mem;
+
+/// In order to avoid too big size of pending events queue,
+/// subordinate state handler removes redundant pending events
+/// and also removes old events if we overflow this limit:
+const MAX_PENDING_EVENTS: usize = 10;
+
+/// [`Subordinate`] is the state of the validator which is not a producer.
+/// It waits for the producer block, the waits for the block computing
+/// and then switches to [`Participant`] state.
+#[derive(Debug, Display)]
+#[display("SUBORDINATE in {:?}", self.state)]
+pub struct Subordinate {
+    ctx: ValidatorContext,
+    producer: Address,
+    block: SimpleBlockData,
+    is_validator: bool,
+    state: State,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum State {
+    WaitingForProducerBlock,
+    WaitingProducerBlockComputed {
+        // TODO +_+_+: change this to producer-block digest when off-chain transactions added
+        block_hash: H256,
+    },
+}
+
+impl StateHandler for Subordinate {
+    fn into_dyn(self: Box<Self>) -> Box<dyn StateHandler> {
+        self
+    }
+
+    fn context(&self) -> &ValidatorContext {
+        &self.ctx
+    }
+
+    fn context_mut(&mut self) -> &mut super::ValidatorContext {
+        &mut self.ctx
+    }
+
+    fn into_context(self: Box<Self>) -> ValidatorContext {
+        self.ctx
+    }
+
+    fn process_block_from_producer(
+        mut self: Box<Self>,
+        block: SignedData<ProducerBlock>,
+    ) -> Result<Box<dyn StateHandler>> {
+        if self.state == State::WaitingForProducerBlock
+            && block.verify_address(self.producer).is_ok()
+            && block.data().block_hash == self.block.hash
```

Comment:
Log trace for the failure might be good, too

Reply:
We have warning in default impl